### PR TITLE
Introduce ComponentIdGenerator

### LINKED ISSUE #40: Render ID in Html
Watching your stream, you mentioned people might have some javascript which depends on the html that webforms is creating. Does that mean that we should render the ID property in the html?

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #40

### CODE REVIEW

File: src/BlazorWebFormsComponents/ComponentIdGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+			}
+
+			generatedId = component.AdditionalAttributes.ContainsKey("name")
+						? component.AdditionalAttributes["name"].ToString()
+						: GenerateId(component);
+
+			return generatedId;
+
+			string GenerateId(BaseWebFormsComponent component)
+				=> $"{component.GetType().Name}_{Guid.NewGuid().ToString()}";
```
Comment (csharpfritz) [inline src/BlazorWebFormsComponents/ComponentIdGenerator.cs]: Web Forms didn't use a GUID for the counter, but rather used an incrementing integer.  Can we use that instead?

**Comment by csharpfritz** [comment]
I think ClientIDMode, as you pointed out, will be necessary to complete this feature.  Not necessary for this PR.

Also:  can we get a unit test to go along with this?

**Comment by hishamco** [comment]
> The Client ID in Web Forms also referenced the parent object ID. Something like: "Container_ListView_1"

I was thought about that at the begging, unfortunately I don't know who access the parent object in the `BaseWebFormComponent`?

> Also: can we get a unit test to go along with this?

Sure

File: src/BlazorWebFormsComponents/ComponentIdGenerator.cs (hunk 1 above)
Comment (hishamco) [reply src/BlazorWebFormsComponents/ComponentIdGenerator.cs]: I already did that, but the issue I faced is I'm unable to know the parent page so the counter increment per page not for all the pages

**Comment by csharpfritz** [comment]
Added #112 to help address this issue.

**Comment by hishamco** [comment]
We have two options either we merge this after adding a unit test or postpone until #112 is fixed

**Comment by egil** [comment]
Have you considered making this a service instead that is injected into each component, and registering it as Scoped?

**Comment by hishamco** [comment]
It could, but I'm not sure if there's a benefit, while it acts as utility class

**Comment by egil** [comment]
It would give you more flexibility in the future, if you wanted to give the generator state, that is scoped to the current session. Perhaps to support numbered IDs.

**Comment by hishamco** [comment]
The intend here is to generate name similar to WebForms, anyhow I may change it after this PR

**Comment by hishamco** [comment]
@csharpfritz do you mean by the container in `Container_ListView_1` the parent control name or the page name

**Comment by hishamco** [comment]
One more thing is there a way to differentiate between the controls in different pages? Coz the incremental counter should be reset on every page

**Comment by csharpfritz** [comment]
> @csharpfritz do you mean by the container in `Container_ListView_1` the parent control name or the page name

This should be the parent control id

**Comment by csharpfritz** [comment]
> One more thing is there a way to differentiate between the controls in different pages? Coz the incremental counter should be reset on every page

This service should be request scoped so that it only is effected by the contents of the current page

**Comment by hishamco** [comment]
> This should be the parent control id

That's what I did in my PR, but it's almost the time `null`

> This service should be request scoped so that it only is effected by the contents of the current page

Could you please elaborate?

**Comment by egil** [comment]
IF you are creating a service to track id's assigned within a single component, you can have the component inherit from OwningComponentBase (see https://chrissainty.com/an-introduction-to-owningcomponentbase/). That way each component gets its own instance. 

If you go in that direction, you could just create a instance of the id generator class in the WebFormsComponentBase class, and not using injection at all (unless you want to be able to change the implementation at app startup).

**Comment by csharpfritz** [comment]
I'm going to close this PR as there has not been any action on it in a few months and I think we need to go in a slightly different direction.

Let's revisit the issue of the ComponentIdGenerator in #40 and discuss strategy

**Comment by hishamco** [comment]
Ya I was quite busy to finalize my MSc thesis, we can revisit the `ComponentIdGenerator` and discuss the strategy as you mentioned

**Pull request closed without merging**